architecture/2025-04-compute-extensibility.md

```diff
@@ -0,0 +1,469 @@
+  name: 'myenv'
+  properties: {
+-    compute: {
+-      kind: 'kubernetes'
+-      namespace: 'default'
+-      }
+-    }
+    recipeConfig: {
+      env {
+        foo: 'bar'
+      }
+    }
+    recipes: {
+      'Applications.Datastores/redisCaches':{
+        default: {
+          templateKind: 'bicep'
+          plainHttp: true
+          templatePath: 'ghcr.io/radius-project/recipes/azure/rediscaches:latest'
+        }
+      }
+    }
+-   extensions: [
+-      {
+-        kind: 'kubernetesMetadata'
+-        labels: {
+-          'team.contact.name': 'frontend'
+-        }
+-      }
+-    ]
+  }
+}
+```
+
+> NOTE: The resource types shown here are new versions, e.g., `environments@2025-05-01-preview`, not changes to the existing versions of the resource types. This allows us to add recipe support to the core types without breaking existing deployments.
+
+The `compute` platform is removed so that we do not have hard-coded support for specific platforms. The functionality enabled by the `compute.identity` property would be implemented via recipes so that we do not need hard-coded support for specific platforms. We could consider adding a recipe to the `environment` type if there are features enabled by `compute` that we could not achieve using recipes on the other core types.
+
+#### Initialize a workspace
+
+As a platform engineer I can use `rad init` (or the equivalent rad `workspace`, `group` and `environment` commands), plus `rad recipe register`, to set up a non-Kubernetes compute platform.
+
+`rad init` would be unchanged. Platform engineers would have to add recipes for each core type and UDT they plan to use.
+
+We could consider adding a flag to `rad init` that would identify a specific set of recipes. For example, `rad init aci` would install a set of default recipes for the ACI platform. However, this work is not in scope for this design.
+
+#### Extend Radius to a new platform by creating new recipes
+
+The `rad recipe register` CLI command would be unchanged, as it already has the ability to associate a recipe to an environment and a resource type, and it supports setting default values for specified recipe parameters for that environment.
+
+```shell
+rad recipe register <recipe name> \
+  --environment <environment name> \
+  --resource-type <core or UDT resource type name> \
+  --parameters throughput=400
+```
+
+#### Set recipes on core types
+
+As a platform engineer I can set recipes on the core resource types of application, container, gateway, and secret store so that I can configure my deployments.
+
+```diff
++resource app 'Applications.Core/applications@2025-05-01-preview' = {
+  name: 'myapp'
+  properties: {
+    environment: environment
+-   extensions: [
+-     {
+-       kind: 'kubernetesNamespace'
+-       namespace: 'myapp'
+-     }
+-     {
+-       kind: 'kubernetesMetadata'
+-       labels: {
+-         'team.contact.name': 'frontend'
+-       }
+-     }
+-   ]
++   recipe: {
```

Comment:
Why would an application need this information?

Reply:
That makes sense. I am removing recipe from applications.